libobs-d3d11/d3d11-shaderprocessor.cpp

```diff
@@ -202,35 +220,104 @@ void ShaderProcessor::BuildSamplers(vector<unique_ptr<ShaderSampler>> &samplers)
 
 void ShaderProcessor::BuildString(string &outputString)
 {
-	stringstream output;
-	output << "static const bool obs_glsl_compile = false;\n\n";
+	stringstream tempOutput, finalOutput;
 
 	cf_token *token = cf_preprocessor_get_tokens(&parser.cfp.pp);
 	while (token->type != CFTOKEN_NONE) {
 		/* cheaply just replace specific tokens */
 		if (strref_cmp(&token->str, "POSITION") == 0)
-			output << "SV_Position";
+			tempOutput << "SV_Position";
 		else if (strref_cmp(&token->str, "TARGET") == 0)
-			output << "SV_Target";
+			tempOutput << "SV_Target";
 		else if (strref_cmp(&token->str, "texture2d") == 0)
-			output << "Texture2D";
+			tempOutput << "Texture2D";
 		else if (strref_cmp(&token->str, "texture3d") == 0)
-			output << "Texture3D";
+			tempOutput << "Texture3D";
 		else if (strref_cmp(&token->str, "texture_cube") == 0)
-			output << "TextureCube";
+			tempOutput << "TextureCube";
 		else if (strref_cmp(&token->str, "texture_rect") == 0)
 			throw "texture_rect is not supported in D3D";
 		else if (strref_cmp(&token->str, "sampler_state") == 0)
-			output << "SamplerState";
+			tempOutput << "SamplerState";
 		else if (strref_cmp(&token->str, "VERTEXID") == 0)
-			output << "SV_VertexID";
-		else
-			output.write(token->str.array, token->str.len);
+			tempOutput << "SV_VertexID";
+		else if (strref_cmp(&token->str, "atomicCounterIncrement") == 0)
+			ReplaceAtomicIncrement(token, tempOutput);
+		else if (!PeekAndSkipAtomicUint(token))
+			tempOutput.write(token->str.array, token->str.len);
+		token++;
+	}
+
+	finalOutput << "static const bool obs_glsl_compile = false;\n\n";
+	if (parser.atomic_counter_next_index > 0)
+		finalOutput
+			<< "RWStructuredBuffer<uint> __uavBuffer : register(u1);\n\n";
+	finalOutput << tempOutput.str();
+
+	outputString = finalOutput.str();
+}
+
+bool ShaderProcessor::SeekUntil(cf_token *&token, const char *str)
```

Comment:
These new functions seem like regular functions, so no need to make them methods?

Reply:
It looked to me like d3d11 is more object-oriented, so I made them members of ShaderProcessor, which is the only place where they are used. They do operate only on an instance of a token and do not require a specific instance of ShaderProcessor, so I've made them static members of the latter. Is there a more proper way to do this?

Reply:
You can see examples of other static functions at the top of the file https://github.com/obsproject/obs-studio/blob/a23434d3a5cb8eb433444819d0c8cd7ee2f3febf/libobs-d3d11/d3d11-shaderprocessor.cpp#L29

Reply:
Gotcha! I added them as static members because this is what I'm used to doing, but I will change things to be consistent with the rest.

Reply:
Changed them to this and they are no longer members of ShaderProcessor:

``` C++

static bool SeekUntil(cf_token *&token, const char *str)
static bool SeekWhile(cf_token *&token, const char *str)
```